### PR TITLE
[POC][sil] Add a dylib that the pass manager in NDEBUG builds can weakly link against and then support verification of SILModules, SILFunctions during optimization.

### PR DESCRIPTION
**Pull request by gottesmm**
The idea is that:

1. We already always ship the frontend option sil-verify-all, it just doesn't do
anything today when NDEBUG is set since we compile out the underlying
verification routines.

2. Given that is true, why not use a weak symbol and the driver's ability to
re-run frontend codegen jobs that crash to instead of just compiling out the
verification to change NDEBUG builds to instead check for the weak pointer and
if it is set, call it for verification. This would then cause -sil-verify-all to
work if the driver when it reran the job to inject the weak symbol.

In the case of the library not being enabled, the cost is just checking if a
function pointer is null which should be cheap. There would be no other
runtime/other cost during normal running meaning that we can use this mechanism
to enable more aggressive verification.

One thing to be aware is that the dylib would be rev-locked to the compiler and
is technically a part of the compiler.

----

NOTE: I still need to make this work for windows and finish the implementation on Linux/test on Linux. This is just a proof of concept.

### CODE REVIEW

File: tools/SourceKit/lib/SwiftLang/CMakeLists.txt [hunk 1]
```diff
@@ -64,3 +64,20 @@ target_link_libraries(SourceKitSwiftLang PRIVATE
   clangAPINotes
   clangBasic)
 add_dependencies(SourceKitSwiftLang clang-tablegen-targets)
+
+if (NOT LLVM_ENABLE_ASSERTIONS)
```
Comment (gottesmm) [inline tools/SourceKit/lib/SwiftLang/CMakeLists.txt:68]: I need to propagate this to the rest of the CMake changes below. But I think in principle this works for Linux/macOS.

**Review by atrick — COMMENTED** [review]
No objections from me. I'm not sure what the ultimate goal is, but this looks better than the current state of affairs. There's no question that the SIL verifier should at least be available in release builds.

**Comment by gottesmm** [comment]
@atrick my main goal here was to make it so that we can rerun with swift-frontend and insert this dylib. In that case, the weak symbol then exists and we actually run the verifier.

That being said, this may be overkill. The only other thing I can think of would be to include the verifier but somehow mark it to be really cold so it is in the same binary when we normally compile.

File: lib/SIL/Verifier/SILVerifierCAPI.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#ifdef SILVERIFIERCAPI_WEAK
+#error "SILVERIFIERCAPI_WEAK defined?!"
+#endif
+
+#ifdef __APPLE__
+#define SILVERIFIERCAPI_WEAK __attribute__((weak_import))
+#elif LINUX
+#define SILVERIFIERCAPI_WEAK __attribute__((weak))
+#else
+#error "Unhandle platform?!"
```
Comment (compnerd) [inline lib/SIL/Verifier/SILVerifierCAPI.h:30]: This 100% does not work on all platforms - Windows does not have the concept of weak linking.  Why not do something like a series of `dlsym` (or the moral equivalent `GetProcAddr`) to get a table of function pointers?

**Comment by compnerd** [comment]
BTW, experiences from clang seem to indicate that including it and marking it as cold might be better.

**Comment by DougGregor** [comment]
I'm not sure this is a good direction for us. At least, we shouldn't start down this part for something as relatively small as the SIL verifier.

**Comment by gottesmm** [comment]
I enabled the verifiers behind -sil-verify-all when asserts are disabled. So we by default don't run them when asserts are disabled, but if -sil-verify-all will run at the beginning of the pipeline, in between passes/etc. In contrast, when asserts are enabled, we run the verifier at strategic points of the pipeline (think after SILGen/before serialization/etc) and -sil-verify-all makes us run it additional times (in between passes/etc)

**Pull request closed without merging**